packages/interaction-tracking/src/InteractionTracking.js

```diff
@@ -0,0 +1,341 @@
+let interactionIDCounter: number = 0;
+let threadIDCounter: number = 0;
+
+// Set of currently tracked interactions.
+// Interactions "stack"–
+// Meaning that newly tracked interactions are appended to the previously active set.
+// When an interaction goes out of scope, the previous set (if any) is restored.
+let interactionsRef: InteractionsRef | null = null;
+
+// Listener(s) to notify when interactions begin and end.
+// Note that subscribers are only supported when enableInteractionTrackingObserver is enabled.
+let subscriberRef: SubscriberRef | null = null;
+
+if (enableInteractionTracking) {
+  interactionsRef = {
+    current: new Set(),
+  };
+  if (enableInteractionTrackingObserver) {
+    subscriberRef = {
+      current: null,
+    };
+  }
+}
+
+// These values are exported for libraries with advanced use cases (i.e. React).
+// They should not typically be accessed directly.
+export {interactionsRef as __interactionsRef, subscriberRef as __subscriberRef};
+
+export function clear(callback: Function): any {
+  if (!enableInteractionTracking) {
+    return callback();
+  }
+
+  const prevInteractions = ((interactionsRef: any): InteractionsRef).current;
+  ((interactionsRef: any): InteractionsRef).current = new Set();
+
+  try {
+    return callback();
+  } finally {
+    ((interactionsRef: any): InteractionsRef).current = prevInteractions;
+  }
+}
+
+export function getCurrent(): Set<Interaction> | null {
+  if (!enableInteractionTracking) {
+    return null;
+  } else {
+    return ((interactionsRef: any): InteractionsRef).current;
+  }
+}
+
+export function getThreadID(): number {
+  return ++threadIDCounter;
+}
+
+export function subscribe(subscriber: Subscriber): void {
+  if (enableInteractionTracking && enableInteractionTrackingObserver) {
+    invariant(
+      ((subscriberRef: any): SubscriberRef).current === null ||
+        ((subscriberRef: any): SubscriberRef).current === subscriber,
+      'Only one interactions subscriber may be registered at a time.',
+    );
+    ((subscriberRef: any): SubscriberRef).current = subscriber;
+  }
+}
+
+export function track(
+  name: string,
+  callback: Function,
+  threadID: number = DEFAULT_THREAD_ID,
+): any {
+  if (!enableInteractionTracking) {
+    return callback();
+  }
+
+  const interaction: Interaction = {
+    id: interactionIDCounter++,
+    name,
+    timestamp: now(),
+  };
+
+  const prevInteractions = ((interactionsRef: any): InteractionsRef).current;
+
+  // Tracked interactions should stack/accumulate.
+  // To do that, clone the current interactions.
+  // The previous set will be restored upon completion.
+  const interactions = new Set(prevInteractions);
+  interactions.add(interaction);
+  ((interactionsRef: any): InteractionsRef).current = interactions;
+
+  if (enableInteractionTrackingObserver) {
+    // Update before calling callback in case it schedules follow-up work.
+    interaction.__count = 1;
+
+    let caughtError;
+    let didCatch = false;
```

Comment:
If you just use nested try/finally where each next level is in the finally, I don’t think you need to use these temporary variables or any catch blocks. That’d probably be better perf wise too.

Reply:
Hm...I could do this, but it would end up throwing the last error, rather than the first. Right?

Reply:
Hm. You're right. It might be worth it though. Just for smaller code and doesn't require the VM to set up registers and initialize these values. It becomes a noop. We don't expect the handlers to ever throw so optimizing for that seems valuable.

Reply:
Okedoke. I'll change it.